proposals/0033-timely-vote-credits.md

```diff
@@ -0,0 +1,389 @@
+of slots to vote on.  When a VoteStateUpdate instruction is processed:
+
+1. For each slot that was in the previous VoteState but are not in the new
+VoteState because they have been rooted in the transition from the old
+VoteState to the new VoteState, credits to award are calculated based on the
+latency that was recorded for them and still available in the old VoteState.
+
+2. For each slot that was in both the previous VoteState and the new
+VoteState, the latency that was previously recorded for that slot is copied
+from the old VoteState to the new VoteState.
+
+3. For each slot that is in the new VoteState but wasn't in the old
+VoteState, the latency value is calculated for this new slot according to
+what slot the vote is for and what slot is in the Clock (i.e. the slot this
+VoteStateUpdate tx landed in) and this latency is stored in VoteState for
+that slot.
+
+The code to handle this is more complex, because VoteStateUpdate may include
+removal of slots that expired as performed by the voting validator, in
+addition to slots that have been rooted and new slots added.  However, the
+assumptions that are needed to handle VoteStateUpdate with timely vote
+credits are already guaranteed by existing VoteStateUpdate correctness
+checking code:
+
+The existing VoteStateUpdate processing code already ensures that (1) only
+roots slots that could actually have been rooted in the transition from the
+old VoteState to the new VoteState, so there is no danger of over-counting
+credits (i.e. imagine that a 'cheating' validator "pretended" that slots
+were rooted by dropping them off of the back of the new VoteState before
+they have actually achieved 32 confirmations; the existing logic prevents
+this).
+
+The existing VoteStateUpdate processing code already ensures that (2) new
+slots included in the new VoteState are only slots after slots that have
+already been voted on in the old VoteState (i.e. can't inject new slots in
+the middle of slots already voted on).
+
+## Impact
+
+1. Any code which deserialized vote account state will require a recompile
+   with the newest Solana SDK in order to be able ot deserialize the new
+   vote account state.
+
+2. There will be a strong anti-incentive for vote lagging.  Validators who
+   currently delay their votes in order to reduce their chances of being
+   locked out from voting, will no longer do so.
+
+3. The Solana CLI will have been updated to show vote latencies for each
+   vote; any context in which the Solana CLI output is observed by users
+   or other tooling will need to adapt to the additonal data being
+   displayed.
+
+## Security Considerations
+
+The are no security considerations; however there are changes to incentive
+structure for voting which may result in differing vote behavior becoming
+more optimal with regards to earning credits.
+
+To maximize vote credits, a validator must vote on all slots that end up
```

Comment:
This paragraph and the 2 under it would be better in the motivation section, not really security considerations.

Reply:
I put that discussion under Security Considerations because are people who believe (and have said so in public discussion) that voting incentives can impact security; so here I lay out the case for why the new voting incentives do not have security implications.